src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs

```diff
@@ -48,14 +50,24 @@ public RemoveDocumentAction(DocumentState document)
                 }
             }
 
-            private class AddDocumentAction : SimpleCompilationTranslationAction<DocumentState>
+            private class AddDocumentsAction : CompilationTranslationAction
             {
-                private static readonly Func<Compilation, DocumentState, CancellationToken, Task<Compilation>> s_action =
-                    async (o, d, c) => o.AddSyntaxTrees(await d.GetSyntaxTreeAsync(c).ConfigureAwait(false));
+                private readonly ImmutableArray<DocumentState> documents;
 
-                public AddDocumentAction(DocumentState document)
-                    : base(document, s_action)
+                public AddDocumentsAction(ImmutableArray<DocumentState> documents)
+                {
+                    this.documents = documents;
+                }
+
+                public override async Task<Compilation> InvokeAsync(Compilation oldCompilation, CancellationToken cancellationToken)
                 {
+                    var syntaxTrees = new List<SyntaxTree>();
+                    foreach (var document in documents)
+                    {
+                        syntaxTrees.Add(await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false));
+                    }
```

Comment:
consider parallelizing for each speedup?

Reply:
I guess just `await Task.WhenAll` would suffice.

Reply:
Let's do this in a separate PR so we can measure benefits/problems of that. This previously wasn't doing that, and although a good suggestion I don't want to couple it with this change.